oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/R2dbcReactiveOAuth2AuthorizedClientService.java

```diff
@@ -0,0 +1,323 @@
+	 * The default is {@link OAuth2AuthorizedClientParametersMapper}.
+	 *
+	 * @param authorizedClientParametersMapper the {@code Function} used for mapping {@link OAuth2AuthorizedClient} and {@link Authentication} to a {@code Map} of {@link String} and {@link Parameter}
+	 */
+	public void setAuthorizedClientParametersMapper(BiFunction<OAuth2AuthorizedClient, Authentication, Map<String, Parameter>> authorizedClientParametersMapper) {
+		Assert.notNull(authorizedClientParametersMapper, "authorizedClientParametersMapper cannot be null");
+		this.authorizedClientParametersMapper = authorizedClientParametersMapper;
+	}
+
+	/**
+	 * Sets the {@link BiFunction} used for mapping the current row in {@code io.r2dbc.spi.Row} to {@link OAuth2AuthorizedClientHolder}.
+	 * The default is {@link OAuth2AuthorizedClientRowMapper}.
+	 *
+	 * @param authorizedClientRowMapper the {@link BiFunction} used for mapping the current row in {@code io.r2dbc.spi.Row} to {@link OAuth2AuthorizedClientHolder}
+	 */
+	public void setAuthorizedClientRowMapper(BiFunction<Row, RowMetadata, OAuth2AuthorizedClientHolder> authorizedClientRowMapper) {
+		Assert.notNull(authorizedClientRowMapper, "authorizedClientRowMapper cannot be null");
+		this.authorizedClientRowMapper = authorizedClientRowMapper;
+	}
+
+	public static class OAuth2AuthorizedClientHolder {
+		private String clientRegistrationId;
+		private OAuth2AccessToken oAuth2AccessToken;
+		private String principalName;
+		private OAuth2RefreshToken refreshToken;
+
+		public OAuth2AuthorizedClientHolder(String clientRegistrationId, OAuth2AccessToken oAuth2AccessToken,
+				String principalName, OAuth2RefreshToken refreshToken) {
+			super();
+			this.clientRegistrationId = clientRegistrationId;
+			this.oAuth2AccessToken = oAuth2AccessToken;
+			this.principalName = principalName;
+			this.refreshToken = refreshToken;
+		}
+
+		public String getClientRegistrationId() {
+			return clientRegistrationId;
+		}
+
+		public OAuth2AccessToken getoAuth2AccessToken() {
+			return oAuth2AccessToken;
+		}
+
+		public String getPrincipalName() {
+			return principalName;
+		}
+
+		OAuth2RefreshToken getRefreshToken() {
+			return refreshToken;
+		}
+	}
+
+	private Mono<OAuth2AuthorizedClient> getOauth2AuthorizedClient(
+			OAuth2AuthorizedClientHolder oAuth2AuthorizedClientHolder) {
+		return this.clientRegistrationRepository
+				.findByRegistrationId(oAuth2AuthorizedClientHolder.getClientRegistrationId())
+				.switchIfEmpty(Mono
+						.error(dataRetrievalFailureException(oAuth2AuthorizedClientHolder.getClientRegistrationId())))
+				.map(clientRegistration -> new OAuth2AuthorizedClient(clientRegistration,
+						oAuth2AuthorizedClientHolder.getPrincipalName(),
+						oAuth2AuthorizedClientHolder.getoAuth2AccessToken(),
+						oAuth2AuthorizedClientHolder.getRefreshToken()));
+	}
+
+	private Mono<Integer> updateAuthorizedClient(OAuth2AuthorizedClient authorizedClient, Authentication principal) {
```

Comment:
Let's move `updateAuthorizedClient` and `insertAuthorizedClient` right under `saveAuthorizedClient`